p4_symbolic/symbolic/values.h

```diff
@@ -0,0 +1,94 @@
+// Copyright 2020 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// This file is responsible for parsing values from the bmv2 json and table
+// entries.
+// It is also responsible for translating any string values to corresponding
+// bitvectors and back, for fields that have the @p4runtime_translation
+// annotation.
+
+#ifndef P4_SYMBOLIC_SYMBOLIC_VALUES_H_
+#define P4_SYMBOLIC_SYMBOLIC_VALUES_H_
+
+#include <string>
+#include <unordered_map>
+
+#include "gutil/status.h"
+#include "p4_pdpi/ir.pb.h"
+#include "z3++.h"
+
+namespace p4_symbolic {
+namespace symbolic {
+namespace values {
+
+// Transforms a hex string literal from bmv2 json to a pdpi::IrValue
+gutil::StatusOr<pdpi::IrValue> ParseIrValue(std::string value);
+
+// Transforms a value read from bmv2 (e.g. hardcoded in the program)
+// to a z3::expr.
+// Essentially, this is just a formatting function that can format ipv4, ipv6
+// hexstrings, and macs as bitvectors in z3's format.
+// This function does not perform any string translation, and instead assumes
+// the values are already in the actual domain of values in the p4 program,
+// because p4 and bmv2 both do not support string translation, it is only used
+// as a logical translation at the boundry between P4RT and bmv2.
+gutil::StatusOr<z3::expr> FormatBmv2Value(const pdpi::IrValue &value);
+
+// This class is responsible for formatting various values into consistent
+// format acceptable by z3.
+// The class transforms pdpi::IrValue instances to z3::expr, it also handles
+// translation (and reverse translation) of P4RT string values that are meant
+// to be translated to bitvectors, because of @p4runtime_translation
+// annotations.
+class ValueFormatter {
```

Comment:
This class does too many things for my taste. The reason that you want a class, is to maintain the invariant of an id allocator. But this class also does all sort of other things (it does formatting, is a bidirectional map for many fields). Can we split it out.

I would create a new class:

class IdAllocator {
  // If an id has already been allocated, return the same value.
  int32 AllocateId(string);
  StatusOr<string> IdToString(int32) const;
}

typedef P4RuntimeTranslationIdAllocators = std::unordered_map<std::string, IdAllocator>